### PR TITLE
Feature link widgets

### PR DESCRIPTION
**Pull request by AdityaKhatri**
Merge Dep: https://github.com/toggle-corp/react-store/pull/339

### CODE REVIEW

File: src/redux/selectors/siloDomainData/analysisFramework.js [hunk 1]
```diff
@@ -2,6 +2,7 @@ import { createSelector } from 'reselect';
 import { afIdFromRoute } from '../domainData';
 
 const emptyObject = {};
+const emptyArray = {};
```
Comment (tnagorra) [inline src/redux/selectors/siloDomainData/analysisFramework.js]: WOW

File: src/widgetComponents/LinkWidgetModal/SupportedWidgets/index.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        supportedWidgets[w.widgetId] !== undefined && w.key !== widgetKey
+    ))
+);
+
+export const getOptionsForSelectedWidget = (selectedWidgetId, widgets) => {
+    const selectedWidget = widgets.find(w => w.key === selectedWidgetId);
+    if (selectedWidget) {
+        return supportedWidgets[selectedWidget.widgetId];
+    }
+    return [];
```
Comment (tnagorra) [inline src/widgetComponents/LinkWidgetModal/SupportedWidgets/index.js]: Prefer not to instantiate new array

File: src/widgetComponents/LinkWidgetModal/SupportedWidgets/organigram.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    const {
+        options,
+        idSelector,
+        labelSelector,
+        childSelector,
+        prefix = '',
+    } = params;
+
+    if (!options || options.length === 0) {
+        return [];
```
Comment (tnagorra) [inline src/widgetComponents/LinkWidgetModal/SupportedWidgets/organigram.js]: Prefer not to instantiate new array

File: src/widgetComponents/LinkWidgetModal/SupportedWidgets/scale.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { mapToList } from '#rsu/common';
+
+const getScaleOptions = (widgetData = {}) => (
+    widgetData.scaleUnits.map(s => ({
```
Comment (tnagorra) [inline src/widgetComponents/LinkWidgetModal/SupportedWidgets/scale.js]: This code looks vulnerable to "Cannot read property map of undefined" attack

File: src/widgetComponents/LinkWidgetModal/index.js [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        data,
+        items,
+        itemValues,
+        itemValue,
+        treeKeySelector,
+        treeLabelSelector,
+        treeNodesSelector,
+    } = params;
+
+    if (!items || items.length === 0) {
```
Comment (tnagorra) [inline src/widgetComponents/LinkWidgetModal/index.js]: The conditionals can be made linear

File: src/widgetComponents/LinkWidgetModal/index.js [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        if (data) {
+            return [{
+                key: treeKeySelector(data),
+                label: treeLabelSelector(data),
+                selected: itemValue.selected,
+            }];
+        }
+        return [];
```
Comment (tnagorra) [inline src/widgetComponents/LinkWidgetModal/index.js]: Prefer not to instantiate new array

File: src/widgetComponents/LinkWidgetModal/index.js [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+            this.selectedWidgetOptions,
+        );
+
+        const widgetData = LinkWidgetModal.getWidgetData(selectedWidget, this.filteredWidgets);
+
+        const treeKeySelector = selectedWidgetOption && selectedWidgetOption.keySelector;
+        const treeLabelSelector = selectedWidgetOption && selectedWidgetOption.labelSelector;
+        const treeNodesSelector = selectedWidgetOption && selectedWidgetOption.nodesSelector;
+
+        const items = selectedWidgetOption ? (selectedWidgetOption.items(widgetData)) : [];
```
Comment (tnagorra) [inline src/widgetComponents/LinkWidgetModal/index.js]: Prefer not to instantiate new array

File: src/widgetComponents/LinkWidgetModal/index.js [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+        const widgetData = LinkWidgetModal.getWidgetData(selectedWidget, this.filteredWidgets);
+
+        const items = selectedWidgetOption ? selectedWidgetOption.items(widgetData) : [];
+        const treeKeySelector = selectedWidgetOption && selectedWidgetOption.keySelector;
+        const treeLabelSelector = selectedWidgetOption && selectedWidgetOption.labelSelector;
+        const treeNodesSelector = selectedWidgetOption && selectedWidgetOption.nodesSelector;
+
+        this.setState({
+            items,
+            itemValues: {},
```
Comment (tnagorra) [inline src/widgetComponents/LinkWidgetModal/index.js]: Prefer not to instantiate new object

File: src/widgetComponents/LinkWidgetModal/index.js [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+            items,
+            itemValues,
+            treeKeySelector,
+            treeLabelSelector,
+            treeNodesSelector,
+            selectedWidget,
+            selectedWidgetItem,
+        } = this.state;
+
+        this.filteredWidgets = getSupportedWidgets(widgets, widgetKey);
```
Comment (thenav56) [inline src/widgetComponents/LinkWidgetModal/index.js:222]: memoize this

File: src/widgetComponents/LinkWidgetModal/index.js [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+        this.selectedWidgetOptions = getOptionsForSelectedWidget(
```
Comment (thenav56) [inline src/widgetComponents/LinkWidgetModal/index.js:223]: memoize this

File: src/widgetComponents/LinkWidgetModal/index.js (hunk 9 above)
Comment (AdityaKhatri) [reply src/widgetComponents/LinkWidgetModal/index.js:222]: Done

File: src/widgetComponents/LinkWidgetModal/index.js (hunk 10 above)
Comment (AdityaKhatri) [reply src/widgetComponents/LinkWidgetModal/index.js:223]: Done

**Pull request merged**